lib/uart.cpp

```diff
@@ -174,23 +170,47 @@ void Uart::write(Uart::Address register_address, uint8_t data) {
         }
         case Uart::Address::PSD://NOTE: Never LSR since that isn't writable
             assert(this->dlab() && "TODO software was mean");//TODO
-            this->m_psd = data;
+            this->regs.m_psd = data;
             break;
         case Uart::Address::MSR://NOTE: Should never be written to by software
             assert(false && "TODO software was mean");//TODO
             break;
         case Uart::Address::SPR: {
-            this->m_spr = data;
+            this->regs.m_spr = data;
             break;
         }
         default: assert(false && "We should never get here!");
     }
 }
 
-bool Uart::interrupt_pending() const {
-    return this->m_isr & 0b1;
+void Uart::update_receive() {
+    uint8_t data;
+    ssize_t bytesRead = ::read(this->receive_file_fd, &data, 1);
+    if(bytesRead > 0){
+            receive_queue.push(data);
+            this->regs.m_isr |= 0b1;
+    }else if(receive_queue.size() == 0){
+        this->regs.m_isr &= ~0b1;
+    }
+}
+
+bool Uart::interrupt_pending() {
+    update_receive();
+    return this->regs.m_isr & 0b1;
 }
 
 bool Uart::dlab() const {
-    return this->m_lcr & (1 << 7);
+    return this->regs.m_lcr & (1 << 7);
+}
+
+void Uart::transmit_thread_function(){
+    std::unique_lock<std::mutex> lock(this->transmit_mutex); 
+    while (!this->kill_transmit_thread){
+        this->transmit_condition_variable.wait(lock);
+        while(this->async_transmit_queue.size() > 0){
+            char data = char(this->async_transmit_queue.front());
+            this->async_transmit_queue.pop();
```

Comment:
In general this could be a synchronization error. Reading the front and popping from the queue should happen in the same lock. Otherwise, after the read and before the pop, another thread could change the queue. In our case, I don't think this would ever be a problem, and I realize that right now our thread safe queue doesn't implement `pop_front` which is what we should be using here.

Reply:
This is a good point. I can try and implement a pop_front to our threadsafe queue.